feat(#46): 이벤트 기반 아키텍처 구현

joel-you opened the pull request:
## 📋 PR 요약

Kafka 기반 이벤트 아키텍처 구현: Producer/Consumer, Retry/DLQ, 모니터링

### 🔗 관련 이슈

- Closes #46

### 📝 변경 사항

- [x] 새로운 기능 추가
- [x] 테스트 코드 추가

### 🛠️ 주요 구현 내용

**1. Kafka 인프라 & 이벤트 모델**
- Docker Compose: Kafka 클러스터 (KRaft)
- 이벤트: EventMetadata, OrderCreated/Updated, PaymentResult, RetryableEvent, FailedEvent

**2. Producer (At-Least-Once)**
- acks=all, idempotence, retries=3
- OrderEventPublisher: 주문 생성/변경 이벤트 발행

**3. Consumer**
- PaymentResultEventConsumer: 결제 결과 처리
- IdempotencyService: Redis 기반 중복 방지 (24h)

**4. Retry & DLQ**
- Exponential Backoff: 2^n * 10초 (최대 3회)
- DLQ: 최종 실패 이벤트 보관

**5. 모니터링**
- KafkaLoggingInterceptor (common 모듈): MDC 분산 추적
- Actuator + Prometheus 메트릭

**6. 통합 테스트**
- Testcontainers 기반 (Kafka/MySQL/Redis)

### 🧪 테스트

  ```bash
  cd infra/docker && docker-compose up -d
  ./gradlew :order-orchestrator:bootRun
  curl -X POST http://localhost:8081/api/v1/orders -H "Content-Type: application/json" -d '{"userId":"user-123","items":[{"productId":"prod-1","quantity":2,"price":10000}]}'
  # Kafka UI: http://localhost:8080
  # Metrics: http://localhost:8081/actuator/prometheus
  ```

  📊 성능

  - 비동기 처리로 응답 시간 개선
  - Kafka 압축 (snappy, ~70%)

  📖 문서

  - KAFKA_SETUP.md, KAFKA_MONITORING.md, DESIGN_PATTERN_COMPARISON.md

  🎯 배포

  - 순서: Kafka → Redis → order-orchestrator
  - application.yml: Kafka, Redis 설정 추가

joel-you committed:
feat(infra): Kafka 클러스터 및 Schema Registry 설정

- Docker Compose에 Schema Registry 및 kafka-init 서비스 추가
- 주문/결제/재시도/DLQ 토픽 자동 생성 설정
  * order.created, order.updated (3 partitions, 7일 retention)
  * payment.result (3 partitions, 7일 retention)
  * retry.events (3 partitions, 1일 retention)
  * dead.letter.queue (1 partition, 30일 retention)
- order-orchestrator application.yml에 Kafka Producer/Consumer 설정 추가
  * Idempotent producer, at-least-once 전달 보장
  * Manual commit, read_committed isolation level
- Testcontainers Kafka 테스트 의존성 추가

joel-you committed:
feat(#46): Kafka Producer 설정 및 이벤트 발행 구현

- KafkaProducerConfig: Idempotent producer 설정
  * acks=all, retries=3로 at-least-once 전달 보장
  * max.in.flight.requests=5로 순서 보장
  * snappy compression으로 네트워크 효율 향상
- KafkaTopics: Topic 상수 정의 (order.created, order.updated 등)
- EventPublisher 인터페이스 및 KafkaEventPublisher 구현
  * 이벤트 validation 및 비동기 발행
  * 성공/실패 로깅 및 예외 처리
- OrderEventPublisher 인터페이스 및 Adapter 구현
  * publishOrderCreated: 주문 생성 이벤트 발행
  * publishOrderUpdated: 주문 상태 변경 이벤트 발행
  * Correlation ID 기반 분산 추적 지원
- Hexagonal Architecture 준수 (Port/Adapter 패턴)

joel-you committed:
feat(order): 주문 생성 및 상태 변경 시 Kafka 이벤트 발행

- CreateOrderUseCase 및 CreateOrderService 구현
  * Order 도메인 생성 및 OrderItem 추가
  * 트랜잭션 커밋 후 OrderCreatedEvent 발행
  * 이벤트 발행 실패 시 OrderEventPublishFailedException
- UpdateOrderStatusUseCase 및 UpdateOrderStatusService 구현
  * 상태 전환 로직: PENDING → CONFIRMED → PAID → PROCESSING → SHIPPED → DELIVERED
  * 트랜잭션 커밋 후 OrderUpdatedEvent 발행
  * 이전 상태 및 변경 사유 포함
- OrderCommandController: REST API 엔드포인트
  * POST /api/v1/orders - 주문 생성
  * PATCH /api/v1/orders/{orderId}/status - 상태 변경
- CreateOrderRequest, UpdateOrderStatusRequest DTO 추가
  * Jakarta Validation 적용
- CreateOrderRequestMapper: MapStruct 매퍼
- Hexagonal Architecture 준수 (UseCase/Port/Adapter)

joel-you committed:
feat(kafka): 결제 결과 이벤트 Consumer 구현

- KafkaConsumerConfig: Consumer 설정
  * Manual commit으로 at-least-once 보장
  * read_committed isolation level
  * JsonDeserializer with PaymentResultEvent
  * Concurrency 3 for parallel processing
- PaymentResultEventConsumer: payment.result 토픽 구독
  * Manual acknowledgment after successful processing
  * Idempotency check using Redis (24h TTL)
  * Exception handling and logging
- ProcessPaymentResultService: 결제 결과 처리 로직
  * SUCCESS → markPaymentCompleted() → PAID
  * FAILED → cancel() → CANCELLED
  * CANCELLED → cancel() → CANCELLED
  * OrderUpdatedEvent 발행 with correlation ID
- IdempotencyService: Redis 기반 중복 방지
  * Event ID 기반 멱등성 보장
- RedisConfig: RedisTemplate 설정

joel-you committed:
feat(kafka): Retry 및 DLQ 에러 처리 메커니즘 구현

- KafkaErrorHandler: 전역 Kafka 에러 핸들러
  * Consumer 처리 실패 시 자동으로 retry topic으로 전송
  * DomainEvent 타입 이벤트 지원
- RetryEventPublisher: Retry/DLQ 발행
  * publishToRetryTopic: 실패 이벤트를 retry.events로 전송
  * publishToDLQ: 최대 재시도 초과 시 dead.letter.queue로 전송
  * 최대 3회 재시도 제한
- RetryEventConsumer: retry.events 토픽 구독
  * 지수 백오프 기반 재시도 (2^n * 10초)
  * nextRetryAt 시간 체크 후 재처리
  * 재시도 카운트 증가 및 추적
  * 최대 재시도 초과 시 자동 DLQ 전송
- Retry Consumer 설정 추가
  * Concurrency 1로 순차 처리
  * Manual commit
- Object 타입 KafkaTemplate 추가
  * RetryableEvent, FailedEvent 직렬화 지원

joel-you committed:
test(#46): Kafka 이벤트 처리 통합 테스트 추가

- Testcontainers를 활용한 Kafka 및 MySQL 통합 테스트 환경 구성
- OrderEventPublisher 통합 테스트: 주문 생성/상태 변경 이벤트 발행 검증
- PaymentResultEventConsumer 통합 테스트: 결제 결과 이벤트 소비 및 처리 검증
- RetryEventConsumer 통합 테스트: Exponential backoff 재시도 로직 검증
- KafkaErrorHandler 통합 테스트: 실패 이벤트의 retry topic 라우팅 검증
- DLQ(Dead Letter Queue) 처리 검증: 최대 재시도 횟수 초과 시 DLQ 이동 확인
- Spring Kafka Test 의존성 추가 및 BuildDependencies에 getter 추가

joel-you committed:
refactor(kafka): MDC 로깅을 공통 Interceptor 패턴으로 개선

- KafkaLoggingInterceptor를 common 모듈로 이동하여 재사용 가능하게 개선
- Lombok @Slf4j 적용으로 코드 간소화
- Consumer 설정에 Interceptor 등록하여 자동 MDC 처리
- PaymentResultEventConsumer, RetryEventConsumer의 수동 MDC 처리 제거
- KafkaMdcHelper 유틸리티 클래스 제거 (Interceptor로 대체)

joel-you committed:
refactor(#46): Kafka Producer 설정을 common 모듈로 통합

- common/KafkaConfig에 성능 튜닝 설정 추가 (압축, 배치, 타임아웃)
- order-orchestrator/KafkaProducerConfig 삭제
- KafkaEventPublisher에서 common의 KafkaTemplate 사용

flab-true commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/adapter/in/kafka/PaymentResultEventConsumer.java:
```diff
@@ -0,0 +1,72 @@
+        try {
+            String eventId = event.getMetadata().getEventId();
+
+            if (idempotencyService.isProcessed(eventId)) {
```
동시 메시지에 대한 중복 처리를 잘 구현해 주셨습니다. 
Redis TTL 은 24시간 설정으로 이해하고 있는데요. Kafka 브로커에서 메시지가 유지되는 설정과 최대한 동일하게 구성해야 되지 않을까 생각합니다. 

Kafka 브로커에서 메시지 유지 기간인 retention.ms 을 확인해 주세요. (`log.retention.ms`)

joel-you replied:
넵 수정했습니다.

joel-you commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/adapter/in/kafka/PaymentResultEventConsumer.java:
```diff
@@ -0,0 +1,72 @@
+                return;
+            }
+
+            processPaymentResultUseCase.processPaymentResult(event);
```
sleep 30초 정도 줘보고 어떻게 동작하는지 봐야 좋을듯

joel-you committed:
Merge remote-tracking branch 'origin/develop' into feature/46-kafka-event-driven

# Conflicts:
#	order-orchestrator/src/main/java/com/msa/commerce/orchestrator/adapter/in/web/OrderCommandController.java

joel-you committed:
refactor(#46): develop merge

joel-you committed:
feat(#46): Transactional Outbox Pattern 구현으로 이벤트 발행 신뢰성 개선

- OutboxEvent 도메인 모델 및 Repository 구현
- OutboxEventRelay 스케줄러로 비동기 이벤트 발행
- 주문 생성/변경 트랜잭션과 이벤트 저장 원자성 보장
- 기존 직접 발행 방식(OrderEventPublisherAdapter) 제거
- IdempotencyService TTL을 Kafka retention과 일치하도록 7일로 수정

joel-you committed:
refactor(#46): Enum 타입 도입 및 Validator 패턴 적용

- String 기반 타입을 Enum으로 변경하여 타입 안정성 향상.
- Builder 패턴 기반 Validator로 검증 로직 개선.
- MapStruct를 통한 Entity-Domain 변환 자동화.
- 사용하지 않는 메서드 및 value 필드 제거.

joel-you committed:
refactor(#46): 마이그레이션 파일 제거

joel-you committed:
fix(#46): Kafka 이벤트 처리 버그 및 보안 개선

- RetryEventConsumer 무한 대기 버그 수정 (nack 추가)
- JsonDeserializer 보안 취약점 해결 (TRUSTED_PACKAGES 제한)
- EventPublishException 활용 및 로깅 개선
- Magic number 제거 및 설정 외부화
- KafkaConsumerConfig 중복 코드 리팩토링

joel-you commented:
#69 PR과 겹치는 것 같은데, 이 PR을 확인해주시면 해당 PR은 Close 하겠습니다

flab-true commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/domain/crossdomain/CrossDomainEvent.java:
```diff
@@ -0,0 +1,149 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@EqualsAndHashCode(of = "id")
+@ToString
+public class CrossDomainEvent {
```
Event 이름이 너무 포괄적인 것 같아요. 실제 비즈니스 이벤트에 대한 이름이면 좋겠습니다.

flab-true commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/adapter/out/crossdomain/CrossDomainEventRelay.java:
```diff
@@ -0,0 +1,78 @@
+                crossDomainEventRepository.save(event);
+
+            } catch (Exception e) {
+                handlePublishFailure(event, e);
```
Error Log는 남기셔도 좋을 것 같습니다.

flab-true commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/adapter/out/crossdomain/CrossDomainEventRelay.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Scheduled(fixedDelay = 1000)
+    @Transactional
+    public void publishPendingEvents() {
```
`PENDING` 상태를 저장하고 Scheduled 기반에 Outbox 패턴으로 처리하는 것으로 이해했습니다. 
해당 @Scheduled는 향후 Spring Batch로 이관되면 좋겠습니다. 발행되는 메시지가 BATCH_SIZE 보다 크다면 이 부분도 이슈가 될 것 같다는 생각입니다.

flab-true commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/adapter/out/persistence/CrossDomainEventJpaRepository.java:
```diff
@@ -0,0 +1,22 @@
+
+public interface CrossDomainEventJpaRepository extends JpaRepository<CrossDomainEventJpaEntity, Long> {
+
+    @Lock(LockModeType.PESSIMISTIC_WRITE)
```
`@Lock(LockModeType.PESSIMISTIC_WRITE)` 왜 사용하셨는지 궁금합니다. 해당 옵션은 테이블 락일까요? Row 락일까요?

flab-true reviewed (commented):
간단한 코멘트 남깁니다.
